src/s3fetchers.ts

```diff
@@ -0,0 +1,113 @@
+import { Readable } from 'stream';
+
+import { S3Client, ListObjectsV2Command, GetObjectCommand } from "@aws-sdk/client-s3";
+
+import { BlockHeight, LakeConfig, Block, Shard, StreamerMessage } from './types';
+import { prependZeroes, parseBody } from './utils';
+
+// Queries the list of the objects in the bucket, grouped by "/" delimiter.
+// Returns the list of blocks that can be fetched
+export async function listBlocks(
+  client: S3Client,
+  bucketName: string,
+  startAfter: BlockHeight,
+): Promise<BlockHeight[]> {
+  try {
+    const data = await client.send(
+        new ListObjectsV2Command({
+          Bucket: bucketName,
+          MaxKeys: 10,
+          Delimiter: "/",
+          StartAfter: prependZeroes(startAfter),
+          RequestPayer: "requester",
+        })
+    );
+    const blockHeights = data.CommonPrefixes.map(p => parseInt(p.Prefix.split('/')[0]));
+    return blockHeights;
+  } catch (err) {
+    console.log("Error", err);
+  }
+}
+
+// By the given block height gets the objects:
+// - block.json
+// - shard_N.json
+// Returns the result in `near_indexer_primitives::StreamerMessage`
+export async function fetchStreamerMessage(
+  client: S3Client,
+  bucketName: string,
+  blockHeight: BlockHeight,
+): Promise<StreamerMessage> {
+  const block = await fetchBlock(client, bucketName, blockHeight);
+  const shards = await fetchShards(client, bucketName, blockHeight, block.chunks.length);
+  const streamerMessage: StreamerMessage = { block, shards };
+  return streamerMessage;
+}
+
+// By the given block height gets the block.json
+// Reads the content of the objects and parses as a JSON.
+async function fetchBlock(
+  client: S3Client,
+  bucketName: string,
+  blockHeight: BlockHeight,
+): Promise<Block> {
+  try {
+    const data = await client.send(
+      new GetObjectCommand({
+        Bucket: bucketName,
+        Key: `${prependZeroes(blockHeight)}/block.json`,
+        RequestPayer: "requester",
+      })
+    );
+    const block: Block = await parseBody<Block>(data.Body as Readable);
+    return block;
+  } catch (err) {
+    console.error(`Failed to fetch ${blockHeight}/block.json. Retrying immediately`, err);
+    return await fetchBlock(client, bucketName, blockHeight);
+  }
+}
+
+// By the given block height gets the shard_N.json files
+// Reads the content of the objects and parses as a JSON.
+async function fetchShards(
+  client: S3Client,
+  bucketName: string,
+  blockHeight: BlockHeight,
+  chunksIncluded: number,
+): Promise<Shard[]> {
+  if (chunksIncluded === 0) return [];
+
+  const shardPromises = [...Array(chunksIncluded).keys()].map(async index => fetchSingleShard(
+    client, bucketName, blockHeight, index
+  ));
+
+  let shards: Shard[] = [];
+
+  for (const shardPromise of shardPromises) {
+    shards.push(await shardPromise);
+  }
+
+  return shards;
```

Comment:
I don't see why this won't work, it seems that we don't need `shards` variable
```
return [...Array(chunksIncluded).keys()].map((index) =>
  fetchSingleShard(client, bucketName, blockHeight, index)
)
```

Reply:
```
Type 'Promise<Shard>[]' is not assignable to type 'Shard[]'.
  Type 'Promise<Shard>' is missing the following properties from type 'Shard': shard_id, receiptExecutionOutcomes, stateChanges
```

Reply:
@khorolets Use `return await Promise.all(...map())`